src/tools/cephfs/cephfs_shell.py

```diff
@@ -0,0 +1,607 @@
+        notation = 'd'
+    for i in mode[-3:]:
+        notation += permission_bits[i]
+    return notation
+
+def get_chunks(file_size):
+    chunk_start = 0
+    chunk_size = 0x20000  # 131072 bytes, default max ssl buffer size
+    while chunk_start + chunk_size < file_size:
+        yield(chunk_start, chunk_size)
+        chunk_start += chunk_size
+    final_chunk_size = file_size - chunk_start
+    yield(chunk_start, final_chunk_size)
+
+def to_bytes(string):
+    return bytes(string, encoding="utf-8")
+
+def list_items(dir_name = ''):
+    if not isinstance(dir_name, bytes):
+        file_name = to_bytes(dir_name)
+    if dir_name == '':
+        d = cephfs.opendir(cephfs.getcwd())
+    else :
+        d = cephfs.opendir(dir_name)
+    dent = cephfs.readdir(d)
+    items = []
+    while dent:
+        if dent.d_name not in [b".", b".."]:
+            items.append(dent)
+        dent = cephfs.readdir(d)
+    cephfs.closedir(d)
+    return items
+
+def print_details(file_name, flag):
+    if not isinstance(file_name, bytes):
+      file_name = to_bytes(file_name)
+    info = cephfs.stat(file_name)
+    print(mode_notation(info.st_mode), info.st_size, info.st_uid, info.st_gid, info.st_mtime, colorama.Fore.BLUE * flag + file_name.decode("utf-8").rsplit('/',1)[1] + '/'*flag + colorama.Style.RESET_ALL * flag,sep='\t')
+
+def word_len(word):
+    """Returns the word lenght, minus any color codes."""
+    if word[0] == '\x1b':
+        return len(word) - 9  
+    return len(word)
+
+def print_list(words, termwidth=79):
+    if not words :
+        return
+    width = max([word_len(word) for word in words])
+    nwords = len(words)
+    ncols = max(1, (termwidth + 1) // (width + 1))
+    nrows = (nwords + ncols - 1) // ncols
+    for row in range(nrows):
+        for i in range(row, nwords, nrows):
+            word = words[i]
+            if word[0] == '\x1b':
+                print('%-*s' % (width + 9, words[i]),
+                           end='\n' if i + nrows >= nwords else ' ')
+            else:
+                print('%-*s' % (width, words[i]),
+                           end='\n' if i + nrows >= nwords else ' ')
+
+def copy_from_local(local_path, remote_path):
+    stdin = -1
+    if local_path == '-' :
+        data = ''.join([line for line in sys.stdin])
+        file_size = len(data)
+    else :
+        file_ = open(local_path, 'rb')
+        stdin = 1
+        file_size = os.path.getsize(local_path)
+        print('File {0}: {1}'.format(local_path, file_size))
+    fd = cephfs.open(to_bytes(remote_path), 'w', 0o666) 
+    if file_size == 0:
+        return  
+    progress = 0
+    while True:
+        data = file_.read(65536)
+        if not data:
+            break
+        wrote = cephfs.write(fd, data, progress) 
+        if wrote < 0:
+            break
+        progress += wrote
+        print('{0} of {1} bytes read ({2}%)'.format(progress, file_size, int(progress / file_size * 100)),  end='\r')
+    cephfs.close(fd)
```

Comment:
Is close(fd) sufficient here? Do you also need to fsync?

Reply:
Why would it be necessary to fsync?

Reply:
Suppose we open the file and get Fb caps. We write a bunch of data to the cache and then ^c out of the shell program. That data may never be written back. It's also possible in the same situation that there could be a writeback error out of the cache, and that may not (or may?) be reported at close. Either way, we're not checking the error on close here so I'm not sure we'd see it and deal with that properly if it does happen.

Reply:
Looking over the code in more detail, I think close basically does flush dirty data and release caps back to the server, so as long as we _know_ that we only have the file open once within the process, we should be ok without fsyncing. Note that this differs from normal POSIX semantics where that is not at all guaranteed.

Reply:
> Suppose we open the file and get Fb caps. We write a bunch of data to the cache and then ^c out of the shell program.

I think we can catch `SIGINT` and shut the mount down properly. I don't like the idea of adding the latency of `fsync` to every file put (especially for recursive put).

Reply:
> Looking over the code in more detail, I think close basically does flush dirty data and release caps back to the server, so as long as we know that we only have the file open once within the process, we should be ok without fsyncing. Note that this differs from normal POSIX semantics where that is not at all guaranteed.

I'm not happy to learn that `ceph_close` performs an `fsync` :(

Reply:
Well, you could just fsync at the end of a recursive put so that might not be quite so bad.

Note too that cached metadata ops could also be a thing here too if you have the right caps. So stuff like a truncate may not always get issued back to the server.

That said, suppose you get something like a SIGKILL or SIGSEGV. You might think that the ops you did first were persisted on storage, but that's not necessarily going to be the case, depending on the activity.

I'm fine with allowing these ops to be done in the client-side cache, but if you're going to do that, then you probably ought to wire up fsync and/or syncfs into this tool to give the caller some way to forcibly flush all pending deltas out. Then be sure to document that anyone using it should issue those calls to ensure that things are persisted correctly.

Reply:
Adding a `sync` call is reasonable.